File: src/store.go
// Atomic episode publication under a contained journal root.
//
// Default layout: YYYY/MM/DD/<episode-id>.md. Non-default classifications
// add reserved worlds/, scopes/, and lanes/ components before the date.
// Each episode is one immutable file per completed turn. Publication is:
// exclusive owner-only temp file in the target directory → write → fsync →
// atomic no-replace hard-link into place → temp unlink → parent directory
// fsync. An existing target is validated by digest: exact duplicate is
// success, mismatch is a typed conflict.
//
// The descent, temp-write, and fsync mechanics publication relies on live in
// corpus.go with the rest of the containment discipline; this file owns the
// publication decision itself.

package autojournal

import (
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"strings"
)

// publishWriteTemp and publishSyncDir are Publish's two durability points —
// the temp-file fsync and the parent-directory fsync — held as package
// variables so the supersede durability witness can observe that the rename
// path exercises exactly the fsyncs the link path does. Production never
// rebinds them.
var (
	publishWriteTemp = writeTemp
	publishSyncDir   = syncDir
	// publishRename is the supersede replacement step, indirected with the
	// fsync points so the witness can assert ordering: the directory fsync
	// is only worth having if it lands after the entry change it makes
	// durable.
	publishRename = func(dir *os.Root, oldname, newname string) error {
		return dir.Rename(oldname, newname)
	}
)

// Published is the result of one Publish call.
type Published struct {
	// Outcome is CapturePublished, CaptureDuplicate, CaptureSuperseded,
	// or CaptureConflict.
	Outcome   CaptureOutcome
	EpisodeID string
	DigestHex string
	// RelPath is the episode path relative to the journal root,
	// slash-joined (the path vocabulary of evidence references).
	RelPath string
	// Content is the rendered episode bytes, so the capture path can
	// index without re-reading the file it just wrote.
	Content []byte
}

// Publish publishes one validated payload into the journal root. The
// world subtree is created on demand with owner-only permissions.
func Publish(root *os.Root, payload *Payload, captureTimeMs uint64) (*Published, error) {
	episodeID := EpisodeID(payload)
	digestHex := PayloadDigestHex(payload)
	content := Render(RenderInput{
		Payload:       payload,
		EpisodeID:     episodeID,
		DigestHex:     digestHex,
		CaptureTimeMs: captureTimeMs,
	})

	components := layoutComponents(payload)
	episodeDir, err := openComponents(root, components)
	if err != nil {
		return nil, err
	}
	defer episodeDir.Close()

	finalName := episodeID + ".md"
	// The temp name embeds the capture time and an attempt counter; a
	// collision (orphan from a crashed writer) retries a fresh name.
	tmpName := ""
	written := false
	for attempt := 0; attempt < 64 && !written; attempt++ {
		tmpName = fmt.Sprintf(".%s.%d.%d.tmp", episodeID, captureTimeMs, attempt)
		err := publishWriteTemp(episodeDir, tmpName, content)
		switch {
		case errors.Is(err, errTempCollision):
			continue
		case err != nil:
			return nil, err
		default:
			written = true
		}
	}
	if !written {
		return nil, fmt.Errorf("temp name: %w: 64 collisions", ErrStoreUnavailable)
	}
	// Whatever happens next, the temp file does not outlive this call.
	defer episodeDir.Remove(tmpName)

	outcome := CapturePublished
	linkErr := episodeDir.Link(tmpName, finalName)
	switch {
	case linkErr == nil:
		// published
	case errors.Is(linkErr, fs.ErrExist):
		outcome, err = classifyExisting(episodeDir, finalName, payload, digestHex)
		if err != nil {
			return nil, err
		}
		if outcome == CaptureSuperseded {
			// The redelivery proved it contains the stored content:
			// replace in place at the episode's own path. The temp is
			// already fsynced, and the directory fsync below makes the
			// replacement entry durable; the superseded bytes are not
			// retained anywhere.
			if err := publishRename(episodeDir, tmpName, finalName); err != nil {
				return nil, corpusError("supersede episode", err)
			}
		}
	default:
		return nil, corpusError("link episode", linkErr)
	}

	// Make the directory entry durable before reporting success.
	if err := publishSyncDir(episodeDir); err != nil {
		return nil, corpusError("sync episode dir", err)
	}

	relPath := strings.Join(append(components, finalName), "/")
	return &Published{
		Outcome:   outcome,
		EpisodeID: episodeID,
		DigestHex: digestHex,
		RelPath:   relPath,
		Content:   content,
	}, nil
}

// classifyExisting decides duplicate, superseded, or conflict for a target
// that already exists at the derived path: it verifies the stored file and,
// when it verifies, tests containment against the incoming payload.
//
// This is where the supersede decision lives, and it deliberately does not
// consult the index. Supersede requires every path-determining field to match, so a
// genuine supersede candidate always lands on the stored episode's own path,
// and the link collision Publish already detects is the only signal needed.
// Supersede therefore works against a missing, stale, or foreign projection —
// which matters, because the alternative would make correcting a settled turn
// depend on index health.
//
// The order of its three tests is part of the contract, because two of them
// can be true of one file and they return different exit codes:
//
//  1. The stored file's *recorded* digest equals digestHex -> duplicate.
//     Deliberately first: an exact redelivery of an episode the owner has
//     since hand-edited stays a duplicate, which is the answer that keeps
//     redelivery idempotent.
//  2. Otherwise verify the stored file. If it verifies and supersedes holds ->
//     superseded.
//  3. Otherwise conflict. This covers a stored file that does not verify,
//     which is never superseded on any path.
//
// The existing file's permissions are repaired to owner-only on the way:
// owner-only is a standing invariant, and a redelivery is a free opportunity
// to fix a file that lost it.
func classifyExisting(dir *os.Root, finalName string, incoming *Payload, digestHex string) (CaptureOutcome, error) {
	f, err := dir.Open(finalName)
	if err != nil {
		return "", corpusError("open existing episode", err)
	}
	defer f.Close()
	if err := f.Chmod(corpusFilePermissions); err != nil {
		return "", corpusError("harden existing episode", err)
	}
	existing, err := io.ReadAll(io.LimitReader(f, MaxEpisodeFileBytes+1))
	if err != nil {
		return "", corpusError("read existing episode", err)
	}
	if len(existing) > MaxEpisodeFileBytes {
		return "", corpusError("read existing episode", fmt.Errorf("exceeds %d bytes", MaxEpisodeFileBytes))
	}
	existingDigest, ok := FrontmatterDigestHex(string(existing))
	if ok && existingDigest == digestHex {
		return CaptureDuplicate, nil
	}
	stored, verifyErr := VerifyEpisode(string(existing))
	if verifyErr == nil && supersedes(incoming, stored) {
		return CaptureSuperseded, nil
	}
	return CaptureConflict, nil
}

// supersedes reports whether incoming is the same turn as stored at a later
// stage of completion. Every field the payload digest covers must be
// identical — world, scope, lane, harness, session id, turn id, event time,
// capture policy, turn outcome, user content — except the two that are
// allowed to grow: stored's assistant result must be a strict prefix of
// incoming's, and stored's tool-name list a prefix of incoming's.
//
// Requiring the event time, scope and lane to match is not belt-and-braces:
// they determine the layout path. A redelivery with a different event time
// shards to another date, publication succeeds rather than colliding, and the
// corpus gains a second file claiming one episode id — the outcome the one-file-per-identity rule rejects,
// reached by accident. A redelivery that shards elsewhere is a conflict.
//
// Anything else is divergence and stays a conflict. Length is not evidence of
// sameness and recency is not evidence of quality; this function is the whole
// of the store's judgment and it is mechanical.
func supersedes(incoming *Payload, stored *VerifiedEpisode) bool {
	if incoming.World != stored.World ||
		incoming.Scope != stored.Scope ||
		incoming.Lane != stored.Lane ||
		incoming.Harness != stored.Harness ||
		incoming.SessionID != stored.SessionID ||
		incoming.TurnID != stored.TurnID ||
		incoming.EventTimeMs != stored.EventTimeMs ||
		incoming.CapturePolicy != stored.CapturePolicy ||
		incoming.TurnOutcome != stored.TurnOutcome ||
		incoming.UserContent != stored.UserContent {
		return false
	}
	if len(incoming.AssistantResult) <= len(stored.AssistantResult) ||
		!strings.HasPrefix(incoming.AssistantResult, stored.AssistantResult) {
		return false
	}
	if len(stored.Tools) > len(incoming.Tools) {
		return false
	}
	for i, tool := range stored.Tools {
		if incoming.Tools[i].Name != tool.Name {
			return false
		}
	}
	return true
}

// Redelivery is a prior capture of the same episode identity.
type Redelivery struct {
	// Outcome is CaptureDuplicate when the stored digest matches the
	// payload, CaptureConflict when the identity was redelivered with
	// different content.
	Outcome CaptureOutcome
	// RelPath is where the existing episode lives.
	RelPath string
}

// CheckRedelivery is the corpus-wide existence check for capture. It exists
// because the store detects a redelivered identity only when it lands on the
// same event-date path, so an identity redelivered with a different event
// time would otherwise shard elsewhere and store twice. The index knows
// every shard, so it answers "does this episode id exist anywhere" — but
// the file it names stays the authority: the outcome is classified from
// that file's own frontmatter, and any index miss, stale row, unreadable
// file, or identity mismatch returns nil so the caller proceeds to
// publish (the store's own same-path check still applies).
//
// It returns nil — proceed to publish — in one more case: the stored episode
// is at the very path this payload derives, and its digest differs. That is
// the supersede candidate, and only Publish's own same-path classification
// can rule on it. Without this, a strict extension would be reported
// conflict before Publish is ever called, because an extended assistant
// result changes the digest by construction; supersede would be unreachable.
// Everything else is unchanged: an exact digest match anywhere is duplicate,
// and a differing digest at a *different* path is a conflict, because supersede
// requires every path-determining field to match.
func CheckRedelivery(root *os.Root, idx *Index, payload *Payload) *Redelivery {
	episodeID := EpisodeID(payload)
	digestHex := PayloadDigestHex(payload)
	row, err := idx.LookupEpisode(episodeID)
	if err != nil || row == nil {
		return nil
	}
	content, err := ReadContained(root, row.RelPath)
	if err != nil {
		return nil
	}
	ep := ParseEpisode(content)
	if ep == nil || ep.EpisodeID != episodeID {
		return nil
	}
	if ep.DigestHex == digestHex {
		return &Redelivery{Outcome: CaptureDuplicate, RelPath: row.RelPath}
	}
	derived := strings.Join(append(layoutComponents(payload), episodeID+".md"), "/")
	if row.RelPath == derived {
		return nil
	}
	return &Redelivery{Outcome: CaptureConflict, RelPath: row.RelPath}
}

// CaptureRequest is one whole capture transaction's input.
//
// Note Defaults' type: config.go's CaptureDefaults (renamed from `type
// Capture`, because src/ is one package and the entry point below
// needs the name Capture).
type CaptureRequest struct {
	RootPath      string
	IndexPath     string
	Raw           RawPayload
	Defaults      CaptureDefaults // owner capture defaults, for world/scope fill
	CaptureTimeMs uint64
}

// CaptureResult is the transaction's typed outcome. Err is nil for every
// success outcome; Detail carries CaptureErrorName(Err) when it is not.
type CaptureResult struct {
	Outcome    CaptureOutcome
	EpisodeID  string
	DigestHex  string
	RelPath    string
	IndexState IndexFreshness
	Err        error
	Detail     string
}

// Capture composes the whole transaction so the imported module form and the
// binary run the same code rather than the same intent: defaults fill,
// Validate, root canonicalization, shared-directory refusal, corpus-wide
// redelivery classification, atomic publication, index update, and the
// index-failure freshness downgrade. Source publication succeeding while
// indexing fails is a success with a downgraded IndexState, never a failure.
//
// The order is part of the contract, not an implementation detail: an
// embedding host that reordered it would report different failures for the
// same input. Shared-directory refusal is decided before the root is opened
// (a refused root is never created), and Validate before either.
func Capture(req CaptureRequest) CaptureResult {
	// Owner-default world/scope fill: a host provides explicit values only
	// when transporting an owner session choice.
	raw := req.Raw
	if raw.World == nil {
		world := req.Defaults.World
		raw.World = &world
	}
	if raw.Scope == nil {
		scope := req.Defaults.Scope
		raw.Scope = &scope
	}
	payload, err := Validate(raw)
	if err != nil {
		return CaptureResult{Outcome: CaptureMalformed, IndexState: IndexNotBuilt,
			Err: err, Detail: CaptureErrorName(err)}
	}

	rootPath := ResolveJournalRoot(req.RootPath)
	if RootInSharedDirectory(rootPath) {
		err := fmt.Errorf("journal root under a shared directory: %w: %w",
			ErrPermissionDenied, ErrSharedDirectory)
		return CaptureResult{Outcome: CapturePermissionDenied, IndexState: IndexNotBuilt,
			Err: err, Detail: CaptureErrorName(err)}
	}
	root, err := OpenJournalRoot(rootPath)
	if err != nil {
		return CaptureResult{Outcome: CaptureUnavailable, IndexState: IndexNotBuilt,
			Err: err, Detail: CaptureErrorName(err)}
	}
	defer root.Close()

	// The index is opened once, with the foreign-root gate, and reused for
	// the redelivery check and the post-publish update. Best-effort
	// throughout: an absent or unhelpful index skips the corpus-wide check
	// (the store's own same-path classification still applies) and
	// downgrades freshness after publication.
	digest := RootDigestHex(rootPath)
	idx, indexErr := OpenIndexHardened(req.IndexPath, &digest)
	if idx != nil {
		defer idx.Close()
		if existing := CheckRedelivery(root, idx, &payload); existing != nil {
			state := IndexFresh
			if existing.Outcome != CaptureDuplicate {
				state = IndexStale
			}
			return CaptureResult{
				Outcome:    existing.Outcome,
				EpisodeID:  EpisodeID(&payload),
				DigestHex:  PayloadDigestHex(&payload),
				RelPath:    existing.RelPath,
				IndexState: state,
			}
		}
	}

	published, err := Publish(root, &payload, req.CaptureTimeMs)
	if err != nil {
		outcome := CaptureUnavailable
		switch {
		case errors.Is(err, ErrContainmentViolation):
			outcome = CaptureInternalError
		case errors.Is(err, ErrPermissionDenied):
			outcome = CapturePermissionDenied
		}
		return CaptureResult{Outcome: outcome, IndexState: IndexNotBuilt,
			Err: err, Detail: CaptureErrorName(err)}
	}

	// Source publication is already durable; the index is best-effort here
	// and repairable via sync, so its failure downgrades freshness only and
	// never changes Outcome.
	indexState := IndexStale
	switch published.Outcome {
	case CaptureConflict:
		// A conflict wrote nothing; there is nothing to index.
	case CapturePublished, CaptureDuplicate, CaptureSuperseded:
		switch {
		case idx == nil && errors.Is(indexErr, ErrForeignIndex):
			indexState = IndexUnavailable
		case idx == nil:
			// Unopenable index: publication stands, projection stays stale.
		default:
			if idx.IndexEpisode(published.RelPath, string(published.Content)) == nil &&
				HardenIndexFiles(req.IndexPath) == nil {
				indexState = IndexFresh
			}
		}
	}
	return CaptureResult{
		Outcome:    published.Outcome,
		EpisodeID:  published.EpisodeID,
		DigestHex:  published.DigestHex,
		RelPath:    published.RelPath,
		IndexState: indexState,
	}
}

File: src/store.ts
// Atomic episode publication under a contained journal root, and the
// whole capture transaction.
//
// Default layout: YYYY/MM/DD/<episode-id>.md. Non-default classifications
// add reserved worlds/, scopes/, and lanes/ components before the date.
// Each episode is one immutable file per completed turn. Publication is:
// exclusive owner-only temp file in the target directory → write → fsync →
// atomic no-replace hard-link into place → temp unlink → parent directory
// fsync. An existing target is classified by recorded digest: exact
// duplicate is success, anything else is a typed conflict. (v1's
// supersede path — in-place replacement of a proven extension — was
// removed in v2: capture fires once per settled turn, so a same-identity
// redelivery with different bytes only ever means divergence.)
//
// The descent, temp-write, and fsync mechanics live in corpus.ts with the
// rest of the containment discipline; this file owns the publication
// decision itself.

import * as fs from "node:fs";
import * as path from "node:path";
import {
  MAX_CONTENT_BYTES,
  MAX_EPISODE_FILE_BYTES,
  validate,
  CaptureError,
  captureErrorName,
  type CaptureOutcome,
  type IndexFreshness,
  type Payload,
  type RawPayload,
  type CaptureErrorCode,
} from "./contracts.ts";
import { episodeId, payloadDigestHex } from "./identity.ts";
import { render } from "./render.ts";
import { frontmatterDigestHex } from "./render.ts";
import type { CaptureDefaults } from "./config.ts";
import {
  descendCreating,
  layoutComponents,
  openJournalRoot,
  rootInSharedDirectory,
  syncDir,
  writeTemp,
  StoreError,
  TempCollisionError,
  type JournalRoot,
} from "./corpus.ts";
import { resolveJournalRoot, rootDigestHex } from "./paths.ts";
import {
  openSnapshot,
  lookupEpisode,
  indexEpisodeIncremental,
  type Snapshot,
} from "./index.ts";
import { readContained } from "./corpus.ts";
import { parseEpisode } from "./episode.ts";

/** The result of one publish call. */
export interface Published {
  outcome: Extract<CaptureOutcome, "published" | "duplicate" | "conflict">;
  episodeId: string;
  digestHex: string;
  /**
   * The episode path relative to the journal root, slash-joined (the path
   * vocabulary of evidence references).
   */
  relPath: string;
  /**
   * The rendered episode content, so the capture path can index without
   * re-reading the file it just wrote.
   */
  content: string;
}

/** Oversize-policy accounting carried alongside a publish. */
export interface DroppedBytes {
  user: number;
  assistant: number;
}

const NO_DROPS: DroppedBytes = { user: 0, assistant: 0 };

/**
 * Publishes one validated payload into the journal root. The world
 * subtree is created on demand with owner-only permissions.
 */
export function publish(
  root: JournalRoot,
  payload: Payload,
  captureTimeMs: number,
  drops: DroppedBytes = NO_DROPS,
): Published {
  const id = episodeId(payload);
  const digestHex = payloadDigestHex(payload);
  const content = render({
    payload,
    episodeId: id,
    digestHex,
    captureTimeMs,
    userDroppedBytes: drops.user,
    assistantDroppedBytes: drops.assistant,
  });
  const contentBytes = Buffer.from(content, "utf8");

  const components = layoutComponents(payload);
  const episodeDir = descendCreating(root, components);

  const finalName = id + ".md";
  // The temp name embeds the capture time and an attempt counter; a
  // collision (orphan from a crashed writer) retries a fresh name.
  let tmpName = "";
  let written = false;
  for (let attempt = 0; attempt < 64 && !written; attempt++) {
    tmpName = `.${id}.${captureTimeMs}.${attempt}.tmp`;
    try {
      writeTemp(episodeDir, tmpName, contentBytes);
      written = true;
    } catch (err) {
      if (err instanceof TempCollisionError) continue;
      throw err;
    }
  }
  if (!written) throw new StoreError("unavailable", "temp name: 64 collisions");

  let outcome: Published["outcome"] = "published";
  try {
    try {
      fs.linkSync(path.join(episodeDir, tmpName), path.join(episodeDir, finalName));
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== "EEXIST") {
        throw classifyIo("link episode", err);
      }
      outcome = classifyExisting(episodeDir, finalName, digestHex);
    }
  } finally {
    // Whatever happened above, the temp file does not outlive this call.
    try {
      fs.rmSync(path.join(episodeDir, tmpName), { force: true });
    } catch {
      // Best-effort: an orphan is invisible to readers.
    }
  }

  // Make the directory entry durable before reporting success.
  syncDir(episodeDir);

  return {
    outcome,
    episodeId: id,
    digestHex,
    relPath: [...components, finalName].join("/"),
    content,
  };
}

function classifyIo(context: string, err: unknown): StoreError {
  const code = (err as NodeJS.ErrnoException).code;
  if (code === "EACCES" || code === "EPERM") return new StoreError("permission_denied", `${context}: ${String(err)}`);
  return new StoreError("unavailable", `${context}: ${String(err)}`);
}

// classifyExisting decides duplicate or conflict for a target that already
// exists at the derived path: the stored file's *recorded* digest equal to
// the incoming digest is a duplicate — deliberately the recorded line, not
// a re-verification, so an exact redelivery of an episode the owner has
// since hand-edited stays a duplicate, which is the answer that keeps
// redelivery idempotent. Anything else — a differing digest, a file with
// no readable digest line — is a conflict.
//
// The existing file's permissions are repaired to owner-only on the way:
// owner-only is a standing invariant, and a redelivery is a free
// opportunity to fix a file that lost it.
function classifyExisting(dirAbs: string, finalName: string, digestHex: string): "duplicate" | "conflict" {
  const finalPath = path.join(dirAbs, finalName);
  let existing: Buffer;
  try {
    // lstat before touching anything: a planted symlink at the final name
    // must not redirect the chmod or the read outside the corpus — the
    // same nofollow discipline every other descent applies.
    if (!fs.lstatSync(finalPath).isFile()) {
      throw new StoreError("containment_violation", "existing episode is not a regular file: " + finalName);
    }
    fs.chmodSync(finalPath, 0o600);
    existing = fs.readFileSync(finalPath);
  } catch (err) {
    if (err instanceof StoreError) throw err;
    throw classifyIo("read existing episode", err);
  }
  if (existing.byteLength > MAX_EPISODE_FILE_BYTES) {
    throw new StoreError("unavailable", `read existing episode: exceeds ${MAX_EPISODE_FILE_BYTES} bytes`);
  }
  const recorded = frontmatterDigestHex(existing.toString("utf8"));
  if (recorded !== null && recorded === digestHex) return "duplicate";
  return "conflict";
}

/**
 * The oversize policy (owner ruling 2026-08-31): a side over the content
 * budget is deterministically tail-truncated to the largest code-point
 * boundary within the budget instead of rejecting the turn, and the
 * dropped byte count is recorded in frontmatter rather than vanishing.
 * Returns the payload to validate plus the per-side accounting.
 */
export function applyOversizePolicy(raw: RawPayload): { raw: RawPayload; drops: DroppedBytes } {
  const [userContent, user] = truncateTail(raw.userContent);
  const [assistantResult, assistant] = truncateTail(raw.assistantResult);
  if (user === 0 && assistant === 0) return { raw, drops: NO_DROPS };
  return { raw: { ...raw, userContent, assistantResult }, drops: { user, assistant } };
}

// truncateTail cuts a string to MAX_CONTENT_BYTES of UTF-8, backing off to
// a code-point boundary, and reports the dropped byte count. Lone
// surrogates would make the byte accounting ill-defined; they are left for
// validate to refuse as InvalidUtf8.
function truncateTail(s: string): [string, number] {
  const total = Buffer.byteLength(s, "utf8");
  if (total <= MAX_CONTENT_BYTES || !s.isWellFormed()) return [s, 0];
  const bytes = Buffer.from(s, "utf8");
  let cut = MAX_CONTENT_BYTES;
  // Back off past UTF-8 continuation bytes (10xxxxxx) to a boundary.
  while (cut > 0 && (bytes[cut] & 0b1100_0000) === 0b1000_0000) cut--;
  const kept = bytes.subarray(0, cut).toString("utf8");
  return [kept, total - cut];
}

/**
 * A prior capture of the same episode identity, found corpus-wide. The
 * projection knows every shard, so it answers "does this episode id exist
 * anywhere" — but the file it names stays the authority: the outcome is
 * classified from that file's own frontmatter, and any index miss, stale
 * row, unreadable file, or identity mismatch returns null so the caller
 * proceeds to publish (the store's own same-path check still applies).
 * A digest mismatch at the very path this payload derives also returns
 * null: only publish's own same-path classification rules there.
 */
export interface Redelivery {
  outcome: "duplicate" | "conflict";
  relPath: string;
}

export function checkRedelivery(root: JournalRoot, snap: Snapshot, payload: Payload): Redelivery | null {
  const id = episodeId(payload);
  const digestHex = payloadDigestHex(payload);
  const row = lookupEpisode(snap, id);
  if (row === null) return null;
  let content: string;
  try {
    content = readContained(root, row.relPath);
  } catch {
    return null;
  }
  const ep = parseEpisode(content);
  if (ep === null || ep.episodeId !== id) return null;
  if (ep.digestHex === digestHex) return { outcome: "duplicate", relPath: row.relPath };
  const derived = [...layoutComponents(payload), id + ".md"].join("/");
  if (row.relPath === derived) return null;
  return { outcome: "conflict", relPath: row.relPath };
}

/**
 * Finds a stored capture of the same turn under a prior capture-policy
 * version. capture_policy participates in episode identity, so a policy
 * bump re-identifies future captures of a turn; an importer redelivering
 * an old session must treat the prior-policy episode as already present
 * or it would double-store every turn captured live before the bump.
 * Returns the stored episode's relPath, or null.
 */
export function findPriorPolicyCapture(
  snap: Snapshot,
  turn: { harness: string; sessionId: string; turnId: string; world: string },
  priorPolicies: string[],
): string | null {
  for (const capturePolicy of priorPolicies) {
    const row = lookupEpisode(snap, episodeId({ ...turn, capturePolicy }));
    if (row !== null) return row.relPath;
  }
  return null;
}

/** One whole capture transaction's input. */
export interface CaptureInput {
  rootPath: string;
  /** Snapshot path; empty skips the projection entirely (tests only). */
  indexPath: string;
  raw: RawPayload;
  /** Owner capture defaults, for world/scope fill. */
  defaults: CaptureDefaults;
  captureTimeMs: number;
}

/**
 * The transaction's typed outcome. detail carries the capture error code
 * for failure outcomes and is empty for every success.
 */
export interface CaptureResult {
  outcome: CaptureOutcome;
  episodeId: string;
  digestHex: string;
  relPath: string;
  indexState: IndexFreshness;
  detail: CaptureErrorCode | "";
  /**
   * True when the refusal was the shared-directory rule specifically, so
   * a renderer can keep its long-standing remediation wording. Wording is
   * rendering; this flag is the typed sentinel.
   */
  sharedDirectory: boolean;
}

/**
 * Composes the whole capture transaction so the extension and the CLI run
 * the same code rather than the same intent: defaults fill, the oversize
 * policy, validate, root canonicalization, shared-directory refusal,
 * corpus-wide redelivery classification, atomic publication, index
 * update, and the index-failure freshness downgrade. Source publication
 * succeeding while indexing fails is a success with a downgraded index
 * state, never a failure. The order is part of the contract: shared-
 * directory refusal is decided before the root is opened (a refused root
 * is never created), and validate before either.
 */
export function capture(input: CaptureInput): CaptureResult {
  const failure = (outcome: CaptureOutcome, detail: CaptureErrorCode, sharedDirectory = false): CaptureResult => ({
    outcome,
    episodeId: "",
    digestHex: "",
    relPath: "",
    indexState: "not_built",
    detail,
    sharedDirectory,
  });

  // Owner-default world/scope fill: a host provides explicit values only
  // when transporting an owner session choice.
  let raw = input.raw;
  if (raw.world === null) raw = { ...raw, world: input.defaults.world };
  if (raw.scope === null) raw = { ...raw, scope: input.defaults.scope };
  const sized = applyOversizePolicy(raw);

  let payload: Payload;
  try {
    payload = validate(sized.raw);
  } catch (err) {
    return failure("malformed", captureErrorName(err));
  }

  const rootPath = resolveJournalRoot(input.rootPath);
  if (rootInSharedDirectory(rootPath)) {
    return failure("permission_denied", "PermissionDenied", true);
  }

  let root: JournalRoot;
  try {
    root = openJournalRoot(rootPath);
  } catch (err) {
    return failure("unavailable", storeErrorCode(err));
  }

  // The snapshot is consulted best-effort: an absent or unhelpful
  // projection skips the corpus-wide check (the store's own same-path
  // classification still applies) and downgrades freshness after
  // publication.
  const digest = rootDigestHex(rootPath);
  const opened = input.indexPath === "" ? null : openSnapshot(input.indexPath, digest);
  if (opened?.kind === "ok") {
    const existing = checkRedelivery(root, opened.snapshot, payload);
    if (existing !== null) {
      return {
        outcome: existing.outcome,
        episodeId: episodeId(payload),
        digestHex: payloadDigestHex(payload),
        relPath: existing.relPath,
        indexState: existing.outcome === "duplicate" ? "fresh" : "stale",
        detail: "",
        sharedDirectory: false,
      };
    }
  }

  let published: Published;
  try {
    published = publish(root, payload, input.captureTimeMs, sized.drops);
  } catch (err) {
    if (err instanceof StoreError) {
      switch (err.code) {
        case "containment_violation":
          return failure("internal_error", "ContainmentViolation");
        case "permission_denied":
          return failure("permission_denied", "PermissionDenied");
        default:
          return failure("unavailable", "Unavailable");
      }
    }
    return failure("unavailable", "Unavailable");
  }

  // Source publication is already durable; the projection update is
  // best-effort and repairable via sync, so its failure downgrades
  // freshness only and never changes the outcome. A conflict wrote
  // nothing; there is nothing to index.
  let indexState: IndexFreshness = "stale";
  if (published.outcome !== "conflict" && input.indexPath !== "") {
    if (opened?.kind === "foreign") {
      indexState = "unavailable";
    } else {
      // ok or not_built alike: the incremental update grows an existing
      // snapshot or births an empty one, matching the v1 engine's
      // create-at-first-capture behavior.
      try {
        const covered = indexEpisodeIncremental(root, input.indexPath, digest, published.relPath, published.content);
        indexState = covered ? "fresh" : "stale";
      } catch {
        indexState = "stale";
      }
    }
  }
  return {
    outcome: published.outcome,
    episodeId: published.episodeId,
    digestHex: published.digestHex,
    relPath: published.relPath,
    indexState,
    detail: "",
    sharedDirectory: false,
  };
}

function storeErrorCode(err: unknown): CaptureErrorCode {
  if (err instanceof StoreError && err.code === "permission_denied") return "PermissionDenied";
  if (err instanceof CaptureError) return err.code;
  return "Unavailable";
}
